commitizen/schemes.py

```diff
@@ -0,0 +1,186 @@
+from __future__ import annotations
+from itertools import zip_longest
+import re
+import sys
+
+from typing import TYPE_CHECKING, ClassVar, Optional, Type
+
+import importlib_metadata as metadata
+from packaging.version import Version as BaseVersion
+from commitizen.defaults import MAJOR, MINOR, PATCH
+
+if TYPE_CHECKING:
+    # TypeAlias is Python 3.10+ but backported in typing-extensions
+    if sys.version_info >= (3, 10):
+        from typing import TypeAlias
+    else:
+        from typing_extensions import TypeAlias
+
+    # Self is Python 3.11+ but backported in typing-extensions
+    if sys.version_info < (3, 11):
+        from typing_extensions import Self
+
+
+DEFAULT_VERSION_PARSER = r"(?P<version>([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?(\w+)?)"
+
+
+class Version(BaseVersion):
```

Comment:
Why was the protocol removed? I think it was a great way to let anyone identify a way to implement their own version schema. Now it seems that we are forcing the use of BaseVersion

Reply:
I removed it for multiple reason:
- we actually rely on almost all `Version` properties
- it felt like duplicating the upstream `Version` class which is already wrapping another one
- `Protocol` only provide type hinting, and while I get the "bring your own" point, when contributing I fill less helped by a protocol than a base class
- it was not documented so I felt it was less useful than the base class (I added some docstrings from the upstream `packaging.version`)
- it was not covering the entire scope (I completed it in the last push)
- there is no syntax to explicitely says a class implement a `Protocol`, unless you inherit it (which doesn't bring much value over inheriting directly the `packaging.version.Version`) but you can't inherit both `Version` (which we requires) and `VersionProtocol`
- `Protocol` are not enforced, they only provide type hinting on duck typing leaving developers with only one choice: have a test suite covering all the expected case (which I doubt given none of the existing plugins is covered) because MyPy won't be able to check that it should and implements the protocol (for the same reason, there is no syntax to express it)
- they don't exists in Python 3.7 while being supported by Commitizen

Given I get the "bring your own" point, I restored it in the last push and completed it (missing properties, docstrings...) and found a way to have proper type semantic with it.